File: src/main/java/frc/chargers/hardware/motorcontrol/ChargerTalonFX.java
package frc.chargers.hardware.motorcontrol;

import com.ctre.phoenix6.BaseStatusSignal;
import com.ctre.phoenix6.StatusSignal;
import com.ctre.phoenix6.configs.TalonFXConfiguration;
import com.ctre.phoenix6.controls.PositionTorqueCurrentFOC;
import com.ctre.phoenix6.controls.PositionVoltage;
import com.ctre.phoenix6.controls.TorqueCurrentFOC;
import com.ctre.phoenix6.controls.VelocityTorqueCurrentFOC;
import com.ctre.phoenix6.controls.VelocityVoltage;
import com.ctre.phoenix6.controls.VoltageOut;
import com.ctre.phoenix6.hardware.TalonFX;
import com.ctre.phoenix6.signals.FeedbackSensorSourceValue;
import com.ctre.phoenix6.signals.NeutralModeValue;
import edu.wpi.first.hal.HAL;
import edu.wpi.first.math.MathUtil;
import edu.wpi.first.units.measure.Angle;
import edu.wpi.first.wpilibj.RobotBase;
import edu.wpi.first.wpilibj.RobotController;
import frc.chargers.hardware.encoders.Encoder;
import frc.chargers.utils.StatusSignalRefresher;
import org.jetbrains.annotations.Nullable;

import java.util.HashSet;
import java.util.Set;

import static edu.wpi.first.math.util.Units.radiansToRotations;
import static edu.wpi.first.math.util.Units.rotationsToRadians;
import static edu.wpi.first.units.Units.Radians;
import static edu.wpi.first.units.Units.RadiansPerSecond;
import static frc.chargers.utils.UtilMethods.tryUntilOk;
import static java.lang.Math.PI;

/**
 * A TalonFX that implements the Motor interface.
 * To access more low-level capabilities of the base api, inherit from this class.
 */
public class ChargerTalonFX implements Motor {
	private static final Set<Integer> usedIds = new HashSet<>();
	public static int getDummyId() {
		int id = 0;
		while (usedIds.contains(id)) {
			id = (int) (Math.random() * 31);
		}
		return id;
	}
	
	protected final TalonFX baseApi;
	private double currentGearRatio = 1.0;
	private boolean useTorqueCurrentControl = false;
	private boolean highFrequencyPosition = false;
	private final StatusSignal<?> positionSignal, velocitySignal, voltageSignal,
		currentSignal, torqueCurrentSignal, supplyCurrentSignal, tempSignal;
	
	protected final VoltageOut setVoltageRequest = new VoltageOut(0);
	protected final PositionVoltage setAngleRequest = new PositionVoltage(0).withSlot(0);
	protected final VelocityVoltage setVelocityRequest = new VelocityVoltage(0).withSlot(1);
	protected final PositionTorqueCurrentFOC setAngleWithTorqueRequest = new PositionTorqueCurrentFOC(0).withSlot(0);
	protected final VelocityTorqueCurrentFOC setVelocityWithTorqueRequest = new VelocityTorqueCurrentFOC(0).withSlot(1);
	protected final TorqueCurrentFOC setTorqueRequest = new TorqueCurrentFOC(0);
	
	private final Encoder encoder = new Encoder() {
		@Override
		public double positionRad() {
			if (highFrequencyPosition) positionSignal.refresh();
			return rotationsToRadians(positionSignal.getValueAsDouble());
		}
		
		@Override
		public double velocityRadPerSec() {
			return rotationsToRadians(velocitySignal.getValueAsDouble());
		}
		
		@Override
		public void setPositionReading(Angle angle) {
			tryUntilOk(baseApi, () -> baseApi.setPosition(angle));
		}
	};
	
	public ChargerTalonFX(int id, boolean optimizeBusUtilization, @Nullable TalonFXConfiguration config) {
		this.baseApi = new TalonFX(id);
		usedIds.add(id);
		this.positionSignal = baseApi.getPosition();
		this.velocitySignal = baseApi.getVelocity();
		this.voltageSignal = baseApi.getMotorVoltage();
		this.currentSignal = baseApi.getStatorCurrent();
		this.supplyCurrentSignal = baseApi.getSupplyCurrent();
		this.tempSignal = baseApi.getDeviceTemp();
		this.torqueCurrentSignal = baseApi.getTorqueCurrent();
		StatusSignal<?>[] allSignals = {
			positionSignal, velocitySignal, voltageSignal, currentSignal,
			supplyCurrentSignal, tempSignal, torqueCurrentSignal
		};
		StatusSignalRefresher.addSignals(allSignals);
		BaseStatusSignal.setUpdateFrequencyForAll(50, allSignals);
		if (optimizeBusUtilization) baseApi.optimizeBusUtilization();
		if (config != null) tryUntilOk(baseApi, () -> baseApi.getConfigurator().apply(config, 0.1));
	}
	
	public ChargerTalonFX withSim(SimDynamics dynamics) {
		if (!RobotBase.isSimulation()) return this;
		var dataInjector = baseApi.getSimState();
		HAL.registerSimPeriodicAfterCallback(() -> {
			dynamics.acceptVolts().accept(MathUtil.clamp(dataInjector.getMotorVoltage(), -12, 12));
			dataInjector.setSupplyVoltage(RobotController.getBatteryVoltage());
			dataInjector.setRawRotorPosition(Radians.of(dynamics.position().getAsDouble() * currentGearRatio));
			dataInjector.setRotorVelocity(RadiansPerSecond.of(dynamics.velocity().getAsDouble() * currentGearRatio));
		});
		return this;
	}
	
	public ChargerTalonFX enablePhoenixPro(boolean useTorqueCurrentControl) {
		setVoltageRequest.EnableFOC = true;
		setAngleRequest.EnableFOC = true;
		setVelocityRequest.EnableFOC = true;
		this.useTorqueCurrentControl = useTorqueCurrentControl;
		return this;
	}
	
	public ChargerTalonFX setPositionUpdateRate(double frequencyHz) {
		positionSignal.setUpdateFrequency(frequencyHz);
		StatusSignalRefresher.remove(positionSignal);
		highFrequencyPosition = true;
		return this;
	}
	
	@Override
	public Encoder encoder() { return encoder; }
	
	@Override
	public double outputVoltage() { return voltageSignal.getValueAsDouble(); }
	
	@Override
	public double statorCurrent() { return currentSignal.getValueAsDouble(); }
	
	@Override
	public double tempCelsius() { return tempSignal.getValueAsDouble(); }
	
	@Override
	public void setVoltage(double volts) {
		baseApi.setControl(setVoltageRequest.withOutput(volts));
	}
	
	@Override
	public void setVelocity(double velocityRadPerSec, double ffVolts) {
		if (useTorqueCurrentControl) {
			setVelocityWithTorqueRequest.Velocity = radiansToRotations(velocityRadPerSec);
			setVelocityWithTorqueRequest.FeedForward = ffVolts;
			baseApi.setControl(setVelocityWithTorqueRequest);
		} else {
			setVelocityRequest.Velocity = radiansToRotations(velocityRadPerSec);
			setVelocityRequest.FeedForward = ffVolts;
			baseApi.setControl(setVelocityRequest);
		}
	}
	
	@Override
	public int id() {
		return baseApi.getDeviceID();
	}
	
	@Override
	public void moveToPosition(double positionRads, double ffVolts) {
		if (useTorqueCurrentControl) {
			setAngleWithTorqueRequest.Position = radiansToRotations(positionRads);
			setAngleWithTorqueRequest.FeedForward = ffVolts;
			baseApi.setControl(setAngleWithTorqueRequest);
		} else {
			setAngleRequest.Position = radiansToRotations(positionRads);
			setAngleRequest.FeedForward = ffVolts;
			baseApi.setControl(setAngleRequest);
		}
	}
	
	@Override
	public void setCoastMode(boolean enabled) {
		baseApi.setNeutralMode(enabled ? NeutralModeValue.Coast : NeutralModeValue.Brake);
	}
	
	@Override
	public void setControlsConfig(ControlsConfig newConfig) {
		var motorConfig = new TalonFXConfiguration();
		tryUntilOk(baseApi, () -> baseApi.getConfigurator().refresh(motorConfig, 0.1));
		if (newConfig.gearRatio() != 1.0) {
			if (motorConfig.Feedback.FeedbackSensorSource == FeedbackSensorSourceValue.RotorSensor) {
				motorConfig.Feedback.SensorToMechanismRatio = newConfig.gearRatio();
			} else {
				motorConfig.Feedback.RotorToSensorRatio = newConfig.gearRatio();
			}
			currentGearRatio = newConfig.gearRatio();
		}
		if (newConfig.positionPID().kP != 0.0) {
			motorConfig.Slot0.kP = newConfig.positionPID().kP * (2 * PI);
			motorConfig.Slot0.kI = newConfig.positionPID().kI * (2 * PI);
			motorConfig.Slot0.kD = newConfig.positionPID().kD * (2 * PI);
			motorConfig.ClosedLoopGeneral.ContinuousWrap = newConfig.continuousInput();
		}
		if (newConfig.velocityPID().kP != 0.0) {
			motorConfig.Slot1.kP = newConfig.velocityPID().kP * (2 * PI);
			motorConfig.Slot1.kI = newConfig.velocityPID().kI * (2 * PI);
			motorConfig.Slot1.kD = newConfig.velocityPID().kD * (2 * PI);
		}
		tryUntilOk(baseApi, () -> baseApi.getConfigurator().apply(motorConfig, 0.1));
	}
	
	@Override
	public double supplyCurrent() { return supplyCurrentSignal.getValueAsDouble(); }
	
	@Override
	public double torqueCurrent() { return torqueCurrentSignal.getValueAsDouble(); }
	
	@Override
	public void setTorqueCurrent(double currentAmps) {
		baseApi.setControl(setTorqueRequest.withOutput(currentAmps));
	}
	
	@Override
	public void close() {
		if (RobotBase.isSimulation()) baseApi.close();
	}
}

File: src/main/kotlin/frc/chargers/hardware/motorcontrol/ChargerTalonFX.kt
@file:Suppress("unused")
package frc.chargers.hardware.motorcontrol

import com.batterystaple.kmeasure.quantities.*
import com.batterystaple.kmeasure.units.*
import com.ctre.phoenix6.StatusCode
import com.ctre.phoenix6.configs.CurrentLimitsConfigs
import com.ctre.phoenix6.configs.MotorOutputConfigs
import com.ctre.phoenix6.configs.TalonFXConfiguration
import com.ctre.phoenix6.controls.*
import com.ctre.phoenix6.hardware.TalonFX
import com.ctre.phoenix6.signals.*
import com.pathplanner.lib.util.PIDConstants
import edu.wpi.first.wpilibj.Alert
import edu.wpi.first.wpilibj.Alert.AlertType
import frc.chargers.framework.UnitTesting
import frc.chargers.hardware.encoders.ChargerCANcoder
import frc.chargers.hardware.encoders.Encoder
import kotlin.math.PI


/**
 * A [TalonFX] motor controller that implements the [Motor] interface
 * and performs periodic self-checking.
 *
 * The [base] property allows for the access of the base [TalonFX] instance.
 */
class ChargerTalonFX(
    val deviceID: Int,
    gearRatio: Double,
    canBus: String? = null,
    factoryDefault: Boolean = true,
    private val faultLogName: String? = null,
    /**
     * Note: this requires the phoenix pro subscription; ask daniel for more details
     */
    fusedCANCoder: ChargerCANcoder? = null
): Motor {
    /**
     * The base [TalonFX] instance.
     */
    val base: TalonFX = if (canBus == null) TalonFX(deviceID) else TalonFX(deviceID, canBus)

    private val config = TalonFXConfiguration()
    private val nonTalonFXFollowers = mutableListOf<Motor>()

    private val positionSignal = base.position
    private val velocitySignal = base.velocity
    private val voltageSignal = base.motorVoltage
    private val currentSignal = base.statorCurrent

    private val voltOutRequest = VoltageOut(0.0).withEnableFOC(true)
    private val setPosRequest = PositionVoltage(0.0).apply {
        Slot = 0
        EnableFOC = true
    }
    private val setVelRequest = VelocityVoltage(0.0).apply {
        Slot = 1
        EnableFOC = true
    }

    private var positionPIDConfigured = false
    private var velocityPIDConfigured = false

    private val errors = mutableListOf<StatusCode>()
    private fun StatusCode.bind() { if (this != StatusCode.OK) errors.add(this) }
    private val faultAlert = Alert("(Motor Fault) $faultLogName", "", AlertType.kError)

    init {
        UnitTesting.addGlobalCloseable(base)
        for (i in 1..4) {
            errors.clear()
            if (!factoryDefault) base.configurator.refresh(config).bind()
            if (fusedCANCoder != null) {
                config.Feedback.FeedbackRemoteSensorID = fusedCANCoder.deviceID
                config.Feedback.FeedbackSensorSource = FeedbackSensorSourceValue.FusedCANcoder
                config.Feedback.RotorToSensorRatio = gearRatio
            } else {
                config.Feedback.SensorToMechanismRatio = gearRatio
            }
            base.configurator.apply(config).bind()
            if (errors.isEmpty()) break
        }
        require(errors.isEmpty()) { "Gear ratio of ChargerTalonFX(id=$deviceID) was not set; critical error." }
    }

    private val faultSignalToMsg = mapOf(
        base.fault_Hardware to "Hardware failure detected",
        base.fault_DeviceTemp to "Device temp exceeded limit",
        base.fault_BootDuringEnable to "Device booted when enabled",
        base.fault_FusedSensorOutOfSync to "Fused CANcoder out of sync",
        base.fault_OverSupplyV to "Voltage exceeded limit",
        base.fault_ProcTemp to "Processor is overheating",
        base.fault_UnlicensedFeatureInUse to "Unlicensed feature in use(git Phoenix Pro Pls)",
        base.fault_UnstableSupplyV to "Supply voltage unstable",
    )

    override fun logFaults() {
        for ((faultSignal, faultMsg) in faultSignalToMsg){
            // != false prevents null
            if (faultSignal.refresh().value != false) {
                faultAlert.setText(faultMsg)
                faultAlert.set(true)
            }
        }
        if (voltageSignal.refresh().status != StatusCode.OK) {
            faultAlert.setText("Device is Unreachable")
            faultAlert.set(true)
        }
    }

    /**
     * The encoder of the TalonFX.
     */
    override val encoder: Encoder = TalonFXEncoderAdapter()
    private inner class TalonFXEncoderAdapter: Encoder {
        override val angularPosition: Angle
            get() = positionSignal.refresh(true).valueAsDouble.ofUnit(rotations)

        override val angularVelocity: AngularVelocity
            get() = velocitySignal.refresh(true).valueAsDouble.ofUnit(rotations/seconds)
    }

    override var voltageOut: Voltage
        get() = voltageSignal.refresh(true).valueAsDouble.ofUnit(volts)
        set(value){
            voltOutRequest.Output = value.siValue
            base.setControl(voltOutRequest)
            nonTalonFXFollowers.forEach{ it.voltageOut = value }
        }

    override val statorCurrent get() = currentSignal.refresh(true).valueAsDouble.ofUnit(amps)

    override val inverted: Boolean get() {
        val motorOutputConfig = MotorOutputConfigs()
        base.configurator.refresh(motorOutputConfig)
        return motorOutputConfig.Inverted == InvertedValue.Clockwise_Positive
    }

    override fun setPositionSetpoint(position: Angle, feedforward: Voltage) {
        if (!positionPIDConfigured) {
            Motor.alertPositionPIDErr()
            return
        } else if (abs(position - this.encoder.angularPosition) < 0.5.degrees) {
            base.setVoltage(0.0)
            return
        }
        setPosRequest.Position = position.inUnit(rotations)
        setPosRequest.FeedForward = feedforward.inUnit(volts)
        base.setControl(setPosRequest)
        nonTalonFXFollowers.forEach{ it.setPositionSetpoint(position, feedforward) }
    }

    override fun setVelocitySetpoint(velocity: AngularVelocity, feedforward: Voltage) {
        if (!velocityPIDConfigured) {
            Motor.alertVelocityPIDErr()
            return
        }
        setVelRequest.Velocity = velocity.inUnit(rotations / seconds)
        setVelRequest.FeedForward = feedforward.inUnit(volts)
        base.setControl(setVelRequest)
        nonTalonFXFollowers.forEach{ it.setVelocitySetpoint(velocity, feedforward) }
    }

    override fun configure(
        inverted: Boolean?,
        brakeWhenIdle: Boolean?,
        rampRate: Time?,
        statorCurrentLimit: Current?,
        followerMotors: List<Motor>,
        positionUpdateRate: Frequency?,
        velocityUpdateRate: Frequency?,
        optimizeUpdateRate: Boolean?,
        currentPosition: Angle?,
        positionPID: PIDConstants?,
        velocityPID: PIDConstants?,
        continuousInput: Boolean?
    ): ChargerTalonFX {
        for (i in 1..4) {
            errors.clear()
            config.MotorOutput.apply {
                when (brakeWhenIdle) {
                    true -> NeutralMode = NeutralModeValue.Brake
                    false -> NeutralMode = NeutralModeValue.Coast
                    null -> {}
                }
                when (inverted) {
                    true -> Inverted = InvertedValue.Clockwise_Positive
                    false -> Inverted = InvertedValue.CounterClockwise_Positive
                    null -> {}
                }
            }
            if (rampRate != null){
                config.OpenLoopRamps.apply {
                    TorqueOpenLoopRampPeriod = rampRate.inUnit(seconds)
                    VoltageOpenLoopRampPeriod = rampRate.inUnit(seconds)
                    DutyCycleOpenLoopRampPeriod = rampRate.inUnit(seconds)
                }
                config.ClosedLoopRamps.apply {
                    TorqueClosedLoopRampPeriod = rampRate.inUnit(seconds)
                    VoltageClosedLoopRampPeriod = rampRate.inUnit(seconds)
                    DutyCycleClosedLoopRampPeriod = rampRate.inUnit(seconds)
                }
            }
            if (statorCurrentLimit != null) {
                config.CurrentLimits.StatorCurrentLimitEnable = true
                config.CurrentLimits.StatorCurrentLimit = statorCurrentLimit.inUnit(amps)
            }
            if (currentPosition != null) base.setPosition(currentPosition.inUnit(rotations)).bind()
            if (continuousInput != null) config.ClosedLoopGeneral.ContinuousWrap = continuousInput
            // 2 * PI makes it so that the PID gains are optimized off of radians and not rotations
            if (positionPID != null) {
                positionPIDConfigured = true
                config.Slot0.apply {
                    kP = positionPID.kP * (2 * PI)
                    kI = positionPID.kI * (2 * PI)
                    kD = positionPID.kD * (2 * PI)
                }
            }
            if (velocityPID != null) {
                velocityPIDConfigured = true
                config.Slot1.apply {
                    kP = velocityPID.kP * (2 * PI)
                    kI = velocityPID.kI * (2 * PI)
                    kD = velocityPID.kD * (2 * PI)
                }
            }
            for (follower in followerMotors){
                follower.configure(
                    positionPID = positionPID,
                    velocityPID = velocityPID,
                    currentPosition = currentPosition
                )
                when (follower) {
                    is ChargerTalonFX -> follower.base.setControl(Follower(this.deviceID, follower.inverted)).bind()
                    else -> nonTalonFXFollowers.add(follower)
                }
            }
            base.configurator.apply(config, 0.1).bind()

            val optimizeBusUtilization = optimizeUpdateRate == true
            if (optimizeBusUtilization) {
                base.optimizeBusUtilization()
                currentSignal.setUpdateFrequency(50.0).bind()
                voltageSignal.setUpdateFrequency(50.0).bind()
            }
            for ((statusSignal, rate) in mapOf(positionSignal to positionUpdateRate, velocitySignal to velocityUpdateRate)){
                if (rate != null) {
                    statusSignal.setUpdateFrequency(rate.inUnit(hertz)).bind()
                } else if (optimizeBusUtilization) {
                    // optimizeBusUtilization will turn the position and velocity signals off by default, so we have to re-enable them
                    statusSignal.setUpdateFrequency(50.0).bind()
                }
            }
            if (errors.isEmpty()) return this
        }
        Alert(
            "${faultLogName ?: "ChargerTalonFX($deviceID)"} could not configure. Errors: $errors",
            AlertType.kError
        ).set(true)
        return this
    }

    fun limitSupplyCurrent(
        limit: Current,
        lowLimit: Current? = null,
        lowLimitActivationTime: Time? = null
    ): ChargerTalonFX {
        val configs = CurrentLimitsConfigs()
        base.configurator.refresh(configs, 0.05)
        configs.apply {
            SupplyCurrentLimitEnable = true
            SupplyCurrentLimit = limit.inUnit(amps)
            if (lowLimit != null) SupplyCurrentLowerLimit = lowLimit.inUnit(amps)
            if (lowLimitActivationTime != null) SupplyCurrentLowerTime = lowLimitActivationTime.inUnit(seconds)
        }
        base.configurator.apply(configs, 0.1)
        return this
    }
}
